modules/tensor_mechanics/src/materials/crystal_plasticity/CrystalPlasticityStressUpdateBase.C

```diff
@@ -0,0 +1,400 @@
+    }
+    if (orthonormal_error)
+      break;
+  }
+
+  fileslipsys.close();
+}
+
+void
+CrystalPlasticityStressUpdateBase::sortCrossSlipFamilies()
+{
+  if (_number_cross_slip_directions == 0)
+  {
+    _cross_slip_familes.resize(0);
+    return;
+  }
+
+  // If cross slip does occur, then set up the system of vectors for the families
+  _cross_slip_familes.resize(_number_cross_slip_directions);
+  // and set the first index of each inner vector
+  for (unsigned int i = 0; i < _number_cross_slip_directions; ++i)
+    _cross_slip_familes[i].resize(1);
+
+  // Sort the index of the slip system based vectors into separte families
+  unsigned int family_counter = 1;
+  _cross_slip_familes[0][0] = 0;
+
+  for (unsigned int i = 1; i < _number_slip_systems; ++i)
+  {
+    for (unsigned int j = 0; j < family_counter; ++j)
+    {
+      // check to see if the slip system direction i matches any of the existing slip directions
+      // First calculate the dot product
+      Real dot_product = 0.0;
+      for (unsigned int k = 0; k < LIBMESH_DIM; ++k)
+      {
+        unsigned int check_family_index = _cross_slip_familes[j][0];
+        dot_product += std::abs(_slip_direction(check_family_index * LIBMESH_DIM + k) -
+                                _slip_direction(i * LIBMESH_DIM + k));
+      }
+      // Then check if the dot product is one, if yes, add to family and break
+      if (MooseUtils::absoluteFuzzyEqual(dot_product, 0.0))
+      {
+        _cross_slip_familes[j].push_back(i);
+        if (_cross_slip_familes[j].size() > _number_cross_slip_planes)
+          mooseError(
+              "Exceeded the number of cross slip planes allowed in a single cross slip family");
+
+        break; // exit the loop over the exisiting cross slip families and move to the next slip
+               // direction
+      }
+      // The slip system in question does not belong to an existing family
+      else if (j == (family_counter - 1) && !MooseUtils::absoluteFuzzyEqual(dot_product, 0.0))
+      {
+        if (family_counter > _number_cross_slip_directions)
+          mooseError("Exceeds the number of cross slip directions specified for this material");
+
+        _cross_slip_familes[family_counter][0] = i;
+        family_counter++;
+        break;
+      }
+    }
+  }
+
+#ifdef DEBUG
+  mooseWarning("Checking the slip system ordering now:");
+  for (unsigned int i = 0; i < _number_cross_slip_directions; ++i)
+  {
+    Moose::out << "In cross slip family " << i << std::endl;
+    for (unsigned int j = 0; j < _number_cross_slip_planes; ++j)
+      Moose::out << " is the slip direction number " << _cross_slip_familes[i][j] << std::endl;
+  }
+#endif
+}
+
+unsigned int
+CrystalPlasticityStressUpdateBase::identifyCrossSlipFamily(const unsigned int index)
+{
+  for (unsigned int i = 0; i < _number_cross_slip_directions; ++i)
+    for (unsigned int j = 0; j < _number_cross_slip_planes; ++j)
+      if (_cross_slip_familes[i][j] == index)
+        return i;
```

Comment:
This is a pretty inefficient way of performing such a lookup. I'd suggest putting the `i`s into an `std::vector<unsigned int>` indexed by `index`. Depending on how often this is called....

Reply:
This is my bad code, not Dewen's. Maybe we could make a separate issue for me to fix up this code in a separate merge request?

As a mitigating factor, this should only get called at the start of the simulation...

Reply:
@sapitts I should have talked to you about this before but I somehow forgot. Glad that Daniel brought this up! 

I actually do not find this called anywhere in the code. Is this for CP classes other than Kalidindi?

Reply:
@dewenyushu this code is used in my BCC iron crystal plasticity classes (where I do need to calculate cross slip), so we could just delete it for now. I um, kind of forgot to take it out when I put in the merge request for the first material-based base class

It is a good example of the kind of stuff I'd like to extrapolate out and have saved in a fancy enum or something. The slip systems for FCC and BCC (and thus the cross slip systems and the twinning systems) are pretty well constant for these materials, so it might be nice for MOOSE to just have those system sets ready to go.

100% a discussion for another issue though

Reply:
I agree it would be better to have the 'slip systems' saved somewhere so that different materials can directly use.  I will keep this in my mind for future discussions